bokehjs/src/lib/models/widgets/progress.ts

```diff
@@ -0,0 +1,175 @@
+    super.connect_signals()
+    const {indeterminate, value, min, max, label, reversed, orientation, disabled, label_location, description} = this.model.properties
+    this.on_change([indeterminate, value, min, max, label, description], () => this._update_value())
+    this.on_change(reversed, () => this._update_reversed())
+    this.on_change(orientation, () => this._update_orientation())
+    this.on_change(disabled, () => this._update_disabled())
+    this.on_change(label_location, () => this._update_label_location())
+  }
+
+  override stylesheets(): StyleSheetLike[] {
+    return [...super.stylesheets(), progress_css.default]
+  }
+
+  override render(): void {
+    super.render()
+
+    this.label_el = div({class: progress_css.label})
+    this.value_el = div({class: progress_css.value})
+    this.bar_el = div({class: progress_css.bar}, this.value_el, this.label_el)
+
+    this._update_value()
+    this._update_disabled()
+    this._update_reversed()
+    this._update_orientation()
+    this._update_label_location()
+
+    this.shadow_el.append(this.bar_el)
+  }
+
+  protected _update_value(): void {
+    const {value, min, max, label} = this.model
+
+    const progress = (clamp(value, min, max) - min)/(max - min)*100
+    const indeterminate = this.model.indeterminate || !isFinite(progress)
+
+    this.class_list.toggle(progress_css.indeterminate, indeterminate)
+    this.value_el.style.setProperty("--progress", `${indeterminate ? 0 : progress}%`)
+
+    this.label_el.textContent = (() => {
+      if (label != null && !indeterminate) {
+        return process_placeholders(label, (_, name) => {
+          return name == "progress" ? `${progress}` : null
+        })
+      } else {
+        return "0%"
+      }
+    })()
+
+    this.bar_el.title = (() => {
+      const {description} = this.model
+      if (description != null && !indeterminate) {
+        return process_placeholders(description, (_, name) => {
+          return name == "progress" ? `${progress}` : null
+        })
+      } else {
+        return ""
+      }
+    })()
+  }
+
+  protected _update_disabled(): void {
+    const {disabled} = this.model
+    this.class_list.toggle(progress_css.disabled, disabled)
+  }
+
+  protected _update_reversed(): void {
+    const {reversed} = this.model
+    this.class_list.toggle(progress_css.reversed, reversed)
+  }
+
+  protected _update_orientation(): void {
+    const {orientation} = this.model
+    this.class_list.toggle(progress_css.horizontal, orientation == "horizontal")
+    this.class_list.toggle(progress_css.vertical, orientation == "vertical")
+  }
+
+  protected _update_label_location(): void {
+    const {label, label_location} = this.model
+    this.label_el.classList.toggle(progress_css.hidden, label == null || label_location == "none")
+  }
+}
+
+export namespace Progress {
+  export type Attrs = p.AttrsOf<Props>
+
+  export type Props = Widget.Props & {
+    indeterminate: p.Property<boolean>
```

Comment:
Instead of a bool why not make this a mode like Angular and Vue:

https://www.creative-tim.com/vuematerial/components/progress-bar

That way we don't prematurely foreclose adding modes like "buffer" and "query" in the future.

Reply:
I'm not sure how well the "reversed" property would play with those potential future modes, though